File: Table_3.py
# Table 3. Longitudinal mixed-effects models.

import warnings

import numpy as np
import pandas as pd
from scipy.stats import norm
import statsmodels.formula.api as smf
from statsmodels.stats.multitest import multipletests

from _common import fmt_p, read_sheet, save_csv

warnings.filterwarnings("ignore", category=UserWarning)


df = read_sheet("Visit_Long_191")
df = df.loc[df["Visit_Order"] <= 7].copy()
df["Time_30d"] = df["Days_From_Baseline"] / 30.0

time_df = (
    df.sort_values(["Study_ID", "Days_From_Baseline", "Visit_Order"])
    .drop_duplicates(["Study_ID", "Days_From_Baseline"], keep="first")
)

outcomes = [
    ("VAS", "VAS"),
    ("CMO, mm", "CMO"),
    ("MMO, mm", "MMO"),
    ("Pain-location burden score", "Pain_Location_Burden_0_6"),
]


def fit_mixedlm(formula, data):
    model = smf.mixedlm(formula, data, groups=data["Study_ID"])
    last_error = None
    for method in ["bfgs", "cg", "powell", "nm"]:
        try:
            fit = model.fit(reml=True, method=method, maxiter=10000, disp=False)
            if fit.converged:
                return fit
        except Exception as exc:
            last_error = exc
    raise RuntimeError(f"MixedLM did not converge: {last_error}")


def estimate(fit, term):
    beta = float(fit.params[term])
    se = float(fit.bse[term])
    p = float(2 * norm.sf(abs(beta / se)))
    return beta, beta - 1.96 * se, beta + 1.96 * se, p


rows = []
for label, outcome in outcomes:
    x = time_df[
        ["Study_ID", "Time_30d", "Baseline_Pain_Group_2L", outcome]
    ].dropna().copy()
    fit_time = fit_mixedlm(
        f"{outcome} ~ Time_30d + C(Baseline_Pain_Group_2L)", x
    )
    tb, tl, th, tp = estimate(fit_time, "Time_30d")

    z = df[
        ["Study_ID", "Visit_Type", "Cycle", "Baseline_Pain_Group_2L", outcome]
    ].dropna().copy()
    fit_visit = fit_mixedlm(
        f'{outcome} ~ C(Visit_Type, Treatment(reference="Injection day")) '
        "+ Cycle + C(Baseline_Pain_Group_2L)",
        z,
    )
    visit_term = next(term for term in fit_visit.params.index if "Visit_Type" in term)
    vb, vl, vh, vp = estimate(fit_visit, visit_term)

    rows.append(
        {
            "Outcome": label,
            "Elapsed-time observations, n": len(x),
            "Elapsed-time patients, n": x["Study_ID"].nunique(),
            "Time_beta": tb,
            "Time_CI_low": tl,
            "Time_CI_high": th,
            "Time_P": tp,
            "Visit-type observations, n": len(z),
            "Visit-type patients, n": z["Study_ID"].nunique(),
            "Visit_beta": vb,
            "Visit_CI_low": vl,
            "Visit_CI_high": vh,
            "Visit_P": vp,
        }
    )

numeric = pd.DataFrame(rows)
numeric["Time_FDR"] = multipletests(numeric["Time_P"], method="fdr_bh")[1]
numeric["Visit_FDR"] = multipletests(numeric["Visit_P"], method="fdr_bh")[1]

formatted = pd.DataFrame(
    {
        "Outcome": numeric["Outcome"],
        "Elapsed-time observations, n": numeric["Elapsed-time observations, n"],
        "Patients, n": numeric["Elapsed-time patients, n"],
        "Per 30-day increase, β (95% CI)": [
            f"{b:.3f} ({lo:.3f} to {hi:.3f})"
            for b, lo, hi in zip(
                numeric["Time_beta"], numeric["Time_CI_low"], numeric["Time_CI_high"]
            )
        ],
        "P value": [fmt_p(p) for p in numeric["Time_P"]],
        "FDR-adjusted P value": [fmt_p(p) for p in numeric["Time_FDR"]],
        "Visit-type observations, n": numeric["Visit-type observations, n"],
        "Visit-type patients, n": numeric["Visit-type patients, n"],
        "Follow-up vs injection-day, β (95% CI)": [
            f"{b:.3f} ({lo:.3f} to {hi:.3f})"
            for b, lo, hi in zip(
                numeric["Visit_beta"], numeric["Visit_CI_low"], numeric["Visit_CI_high"]
            )
        ],
        "Visit-type P value": [fmt_p(p) for p in numeric["Visit_P"]],
        "Visit-type FDR-adjusted P value": [fmt_p(p) for p in numeric["Visit_FDR"]],
    }
)

save_csv(numeric, "Table_3_numeric_Python.csv")
save_csv(formatted, "Table_3.csv")
print(formatted.to_string(index=False))

File: Table_3.R
# Table 3. Longitudinal mixed-effects models.

suppressPackageStartupMessages({
  library(readxl)
  library(dplyr)
  library(lme4)
})

source(file.path(
  dirname(sub("^--file=", "", commandArgs(FALSE)[grepl("^--file=", commandArgs(FALSE))][1])),
  "_common.R"
))

dat <- read_excel(resolve_data_path(), sheet = "Visit_Long_191") %>%
  filter(Visit_Order <= 7) %>%
  mutate(
    Study_ID = factor(Study_ID),
    Visit_Type = factor(Visit_Type, levels = c("Injection day", "Follow-up")),
    Baseline_Pain_Group_2L = factor(Baseline_Pain_Group_2L, levels = c(MILD, MODSEV)),
    Time_30d = Days_From_Baseline / 30
  )

# One record per patient-day for the elapsed-time model.
same_day_qc <- dat %>%
  count(Study_ID, Days_From_Baseline, name = "n_records") %>%
  filter(n_records > 1)

write.csv(
  same_day_qc,
  file.path(output_dir(), "Table_3_same_day_QC.csv"),
  row.names = FALSE,
  fileEncoding = "UTF-8"
)

dat_time <- dat %>%
  arrange(Study_ID, Days_From_Baseline, Visit_Order) %>%
  group_by(Study_ID, Days_From_Baseline) %>%
  slice(1) %>%
  ungroup()

outcomes <- c("VAS", "CMO", "MMO", "Pain_Location_Burden_0_6")
labels <- c(
  VAS = "VAS",
  CMO = "CMO, mm",
  MMO = "MMO, mm",
  Pain_Location_Burden_0_6 = "Pain-location burden score"
)

fit_elapsed <- function(outcome) {
  fit <- lmer(
    as.formula(paste0(
      outcome,
      " ~ Time_30d + Baseline_Pain_Group_2L + (1 | Study_ID)"
    )),
    data = dat_time,
    REML = TRUE,
    na.action = na.omit
  )

  co <- summary(fit)$coefficients
  beta <- co["Time_30d", "Estimate"]
  se <- co["Time_30d", "Std. Error"]
  mf <- model.frame(fit)

  data.frame(
    Outcome = outcome,
    Observations = nobs(fit),
    Patients = n_distinct(mf$Study_ID),
    Beta = beta,
    CI_low = beta - 1.96 * se,
    CI_high = beta + 1.96 * se,
    P = 2 * pnorm(abs(beta / se), lower.tail = FALSE),
    Singular = isSingular(fit, tol = 1e-4),
    stringsAsFactors = FALSE
  )
}

fit_visit <- function(outcome) {
  fit <- lmer(
    as.formula(paste0(
      outcome,
      " ~ Visit_Type + Cycle + Baseline_Pain_Group_2L + (1 | Study_ID)"
    )),
    data = dat,
    REML = TRUE,
    na.action = na.omit
  )

  co <- summary(fit)$coefficients
  term <- "Visit_TypeFollow-up"
  beta <- co[term, "Estimate"]
  se <- co[term, "Std. Error"]
  mf <- model.frame(fit)

  data.frame(
    Outcome = outcome,
    Observations = nobs(fit),
    Patients = n_distinct(mf$Study_ID),
    Beta = beta,
    CI_low = beta - 1.96 * se,
    CI_high = beta + 1.96 * se,
    P = 2 * pnorm(abs(beta / se), lower.tail = FALSE),
    Singular = isSingular(fit, tol = 1e-4),
    stringsAsFactors = FALSE
  )
}

elapsed <- bind_rows(lapply(outcomes, fit_elapsed)) %>%
  mutate(FDR = p.adjust(P, method = "BH"))

visit <- bind_rows(lapply(outcomes, fit_visit)) %>%
  mutate(FDR = p.adjust(P, method = "BH"))

numeric_table <- left_join(
  elapsed,
  visit,
  by = "Outcome",
  suffix = c("_Time", "_Visit")
)

numeric_table$Outcome <- unname(labels[numeric_table$Outcome])

fmt_ci <- function(beta, low, high) {
  sprintf("%.3f (%.3f to %.3f)", beta, low, high)
}

table3 <- numeric_table %>%
  transmute(
    Outcome,
    `Elapsed-time observations, n` = Observations_Time,
    `Patients, n` = Patients_Time,
    `Per 30-day increase, β (95% CI)` = fmt_ci(Beta_Time, CI_low_Time, CI_high_Time),
    `P value` = sapply(P_Time, fmt_p),
    `FDR-adjusted P value` = sapply(FDR_Time, fmt_p),
    `Visit-type observations, n` = Observations_Visit,
    `Visit-type patients, n` = Patients_Visit,
    `Follow-up vs injection-day, β (95% CI)` = fmt_ci(Beta_Visit, CI_low_Visit, CI_high_Visit),
    `Visit-type P value` = sapply(P_Visit, fmt_p),
    `Visit-type FDR-adjusted P value` = sapply(FDR_Visit, fmt_p)
  )

write.csv(
  numeric_table,
  file.path(output_dir(), "Table_3_numeric_R.csv"),
  row.names = FALSE,
  fileEncoding = "UTF-8"
)
write.csv(
  table3,
  file.path(output_dir(), "Table_3.csv"),
  row.names = FALSE,
  fileEncoding = "UTF-8"
)

print(table3)
